### ISSUE TITLE
Question; Shouldn't MimePart.IsAttachment return true also for linked resources ...?

### ISSUE THREAD

**Issue by polterguy** [jstedfast/MimeKit#202]
Howdy, I am creating this message through BodyBuilder with an inline attachment. Headers for attachment becomes something like this;
Content-Type: image/png; name=marvin-headshot.png
Content-Disposition: inline; filename=marvin-headshot.png
Content-Location: marvin-headshot.png
Content-Transfer-Encoding: base64

When I retrieve this message, I realise I cannot use "IsAttachment" from MimePart to figure out if it's an attachment or not ...

I know that according to "Content-Location" header this isn't "really" an attachment, however, logically in client code, it needs to be handled as an attachment, which normally/often means storing it to some place on disc I guess ...

So basically; How do I figure out the difference between an attachment and text content, when I (obviously) cannot use the MimePart.IsAttachment property ...?

I guess I could figure out such a thing by looking at Content-Type, and realise it's impossible to handle as anything but an attachment, being "image/png" that is. But in theory, one might imagine inline attachments that are of type "text/whatever" something ...?

For instance, a theoretical use case for such a thing, would be to create an inline attachment for CSS, which would create Content-Type of "text/css", meaning MimePart would be TextPart, while also being an inline attachment, for linking into some CSS declaration of some piece of HTML in another MimePart. Or for that matter; JavaScript ...? (With all the security issues that would raise put aside for the sake of the argument)

But maybe that's stretching it ...

Anyway, howto figure out if something is an attachment or not ...?

Should I just check the MimePart.Filename property to see if something is an attachment or not ...?
Would that be the (logical) correct way of figuring out if something is an attachment ...?

**Comment by jstedfast** [comment]
It's a subjective issue. Most of the time, when you have a `Content-Disposition: inline` header, you don't want to treat it as an attachment. You only want to treat it as an attachment if it has `Content-Disposition: attachment`.

It's one of those places where you can never satisfy everyone.

Since, in general, inline images are displayed inline as part of the message body, most mail clients don't include them in the list of "attachments" that they show the user, so I went with the same semantic.

You can certainly check that the FileName is non-null to decide if it is an attachment or not. There's definitely a number of people that do that...

Some mail clients use a more complicated heuristic, which is to say that as they render the HTML message body, they track which `inline` images were included, and any that were not, are also dumped into the 'attachment' list.

To implement this sort of heuristic, you could use the MimeVisitor example code at https://github.com/jstedfast/MailKit/blob/master/Documentation/Examples/MimeVisitorExamples.cs

Basically, once you "visit" the MimeMessage, you can access `visitor.Attachments` to get all of the "not used while rendering this message" list of MimeParts, which can probably safely be assumed to be attachments.

Hope that helps.

**Comment by polterguy** [comment]
Thx, I will look into this :)
